python/cuml/dask/common/input_utils.py

```diff
@@ -0,0 +1,314 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import cudf
+import cupy as cp
+import dask.array as da
+
+from cuml.dask.common.dask_df_utils import to_dask_cudf
+from cuml.utils.memory_utils import with_cupy_rmm
+
+from collections import OrderedDict
+from cudf.core import DataFrame
+from dask_cudf.core import DataFrame as dcDataFrame
+
+from dask import delayed
+from dask.distributed import wait
+from dask.distributed import default_client
+from tornado import gen
+from toolz import first
+from collections.abc import Iterable
+from dask.array.core import Array as daskArray
+from uuid import uuid1
+
+from collections import defaultdict
+from functools import reduce
+
+import dask.dataframe as dd
+
+
+class MGData:
+
+    def __init__(self, gpu_futures=None, worker_to_parts=None, workers=None,
```

Comment:
I like the static methods a lot, but I'm struggling to wrap my head around what an instance of MGData represents, as it seems to hold a lot different important items. Is it a wrapper that encapsulates a single dataset - basically the equivalent of CumlArray for Dask? Or an MGDataManager that manages the flow for one computation? It seems to hold a lot more info than just a CumlArray-equivalent.

Maybe the answer is that users should always use a static factory method and never instantiate this themselves?

Reply:
The answer is indeed the last one, the static factory methods are the API to use, and will allow us to extend this abstraction for future needs. 

After sleeping on it, how about DistributedDataHandler for the class name?  A little verbose, but it might be clearer what is its purpose: to handle and abstract away all the needed operations for handling distributed data: getting futures, sizes, and whatnot